# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/WasmAnyRef.cpp
# Commit: 72bf9c871ded
# Full Hash: 72bf9c871ded006f56f09b24d0723365516445bb
# Author: Ryan Hunt <rhunt@eqrion.net>
# Date: 2023-08-05 09:19:01
# Regressor Bug: 1692065
# File Overlap Count: 5
# Description:
#   Bug 1692065 - wasm: Add tagging for 31-bit integers and i31ref. r=yury
#   
#   This commit adds a tag 0x1 for 31-bit integers in AnyRef. Any JS value
#   that is losslessly representable (without coercion) in this range is
#   stored without boxing. This is chosen to line up exactly with how the
# ==============================================================================

diff -r 28feda14fdec -r 72bf9c871ded js/src/wasm/WasmAnyRef.cpp
--- a/js/src/wasm/WasmAnyRef.cpp	Fri Aug 04 14:06:50 2023 +0000
+++ b/js/src/wasm/WasmAnyRef.cpp	Fri Aug 04 14:06:50 2023 +0000
@@ -76,6 +76,21 @@
     return true;
   }
 
+  if (value.isInt32() && !int32NeedsBoxing(value.toInt32())) {
+    result.set(AnyRef::fromInt32(value.toInt32()));
+    return true;
+  }
+
+  if (value.isDouble()) {
+    double doubleValue = value.toDouble();
+    int32_t intValue;
+    if (mozilla::NumberIsInt32(doubleValue, &intValue) &&
+        !int32NeedsBoxing(intValue)) {
+      result.set(AnyRef::fromInt32(intValue));
+      return true;
+    }
+  }
+
   JSObject* box = AnyRef::boxValue(cx, value);
   if (!box) {
     return false;
@@ -97,6 +112,8 @@
     value.setNull();
   } else if (isJSString()) {
     value.setString(toJSString());
+  } else if (isI31()) {
+    value.setInt32(toI31());
   } else {
     JSObject& obj = toJSObject();
     if (obj.is<WasmValueBox>()) {